src/ophys_etl/modules/neuropil_correction/utils.py

```diff
@@ -0,0 +1,404 @@
+
+    return mat_dict
+
+
+def ab_from_diagonals(mat_dict: dict) -> np.ndarray:
+    """Constructs value for scipy.linalg.solve_banded
+
+    Parameters
+    ----------
+    mat_dict: dict
+        dictionary of diagonals keyed by offsets
+
+    Returns
+    -------
+    ab: np.ndarray
+        value for scipy.linalg.solve_banded
+    """
+    offsets = list(mat_dict.keys())
+    ll = -np.min(offsets)
+    u = np.max(offsets)
+
+    T = mat_dict[offsets[0]].shape[0]
+
+    ab = np.zeros([ll + u + 1, T])
+
+    for o in offsets:
+        index = u - o
+        ab[index] = mat_dict[o]
+
+    return ab
+
+
+def error_calc(
+    F_M: np.ndarray,
+    F_N: np.ndarray,
+    F_C: np.ndarray,
+    r: float,
+) -> float:
+    """Calculate root mean square error between corrected trace
+    and roi trace with subtracted neuropil contamination.
+
+    Parameters
+    -----------
+    F_M_array: np.ndarray
+        ROI trace
+    F_N: np.ndarray
+        neuropil trace
+    F_C: np.ndarray
+        neuropil corrected trace
+    r: float
+        contamination ratio
+
+    Returns
+    --------
+    er: float
+        RMSE
+    """
+    er = np.sqrt(np.mean(np.square(F_C - (F_M - r * F_N)))) / np.mean(F_M)
+
+    return er
+
+
+def ab_from_T(T: int, lam: float, dt: float):
+    # using csr because multiplication is fast
+    Ls = -sparse.eye(T - 1, T, format="csr") + sparse.eye(
+        T - 1, T, 1, format="csr"
+    )
+    Ls /= dt
+    Ls2 = Ls.T.dot(Ls)
+
+    M = sparse.eye(T) + lam * Ls2
+    mat_dict = get_diagonals_from_sparse(M)
+    ab = ab_from_diagonals(mat_dict)
+
+    return ab
+
+
+def fill_unconverged_r(
+    corrected_neuropil_traces: np.ndarray,
+    roi_traces: np.ndarray,
+    neuropil_traces: np.ndarray,
+    r_array: np.ndarray,
+    flag_threshold: float = 1.0,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """R values that are unconverged as defined by values that exceed 1.0
+    are filled in with the mean value of all other cells from the same
+    experiment. The corrected fluorescence trace is recalculated with the
+    adjusted R value.
+    """
+    flagged_mask = r_array > flag_threshold
+    fill_r_val = r_array[(~flagged_mask) & (r_array > 0)].mean()
+    r_array[flagged_mask] = fill_r_val
+    corrected_neuropil_traces[flagged_mask] = (
+        roi_traces[flagged_mask] - fill_r_val * neuropil_traces[flagged_mask]
+    )
+    rmse = []
+    for F_M, F_N, r in zip(roi_traces, neuropil_traces, r_array):
+        ns = NeuropilSubtract()
+        ns.set_F(F_M, F_N)
+        rmse.append(ns.estimate_error(r))
+    return corrected_neuropil_traces, r_array, np.array(rmse)
+
+
+class NeuropilSubtract(object):
+    """Class to jointly estimate the corrected fluorescence
+    trace (F_C) and the R contamination value given
+    a neuropil trace and ROI trace
+
+    The estimation is performed with a gradient descent to
+    minimize a loss function:
+
+    E = (F_C - (F_M - r*F_n))**2 + lam*dt*F_C
```

Comment:
Can you define F_M and F_n in the docstring here? I assume they are `measured` and `neuropi` traces but it would be good to make it clear (and confirm my assumption).

Reply:
I just added the init parameters there. 
The F_M and F_N are in the docstring under set_F

Reply:
As this is likely the first place you will see them, you should define all the terms that make up the function above.